crates/bevy_ecs/src/archetype.rs

```diff
@@ -453,38 +449,33 @@ impl Archetypes {
         table_components: Vec<ComponentId>,
         sparse_set_components: Vec<ComponentId>,
     ) -> ArchetypeId {
-        let table_components = table_components.into_boxed_slice();
-        let sparse_set_components = sparse_set_components.into_boxed_slice();
         let archetype_identity = ArchetypeIdentity {
-            sparse_set_components: sparse_set_components.clone(),
-            table_components: table_components.clone(),
+            sparse_set_components: sparse_set_components.clone().into_boxed_slice(),
+            table_components: table_components.clone().into_boxed_slice(),
         };
 
         let archetypes = &mut self.archetypes;
         let archetype_component_count = &mut self.archetype_component_count;
-        let mut next_archetype_component_id = move || {
-            let id = ArchetypeComponentId(*archetype_component_count);
-            *archetype_component_count += 1;
-            id
-        };
         *self
             .archetype_ids
             .entry(archetype_identity)
```

Comment:
Not something that needs to be done here, but I think we can completely eliminate the redundant clones / allocations here related to ArchetypeIdentity:

1. Use Hashbrown's raw_entry API to remove the need for an "owning" ArchetypeIdentity when the archetype already exists
2. When the archetype does not exist, create the Archetype first using a non-owning iterator over sparse_set_components and table_components, then use the same allocated lists to construct the final owning ArchetypeIdentity.